### PR TITLE
Ilina/fastsync concurrency

### PR DESCRIPTION
**Pull request by istoilkovska**
Modeling of bounded and blocking buffers.

### CODE REVIEW

File: spec/fastsync/fastsync_concurrency_with_peers.tla [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+    \/ SendPeerMsgs
+    \* do nothing 
+    \/ /\ UNCHANGED buffers
+       /\ UNCHANGED peer
+
+Buffer ==
+    \/ /\ ~Ready(receiveDemux)
+       /\ ~Full(receiveDemux)
+       /\ receiveDemux' = BufRcv(receiveDemux)
+       /\ UNCHANGED <<demuxSend, demuxProcess, processDemux, demuxSchedule, scheduleDemux, peerIncoming, peerOutgoing, peerTurn>>
```
Comment (konnov) [inline spec/fastsync/fastsync_concurrency_with_peers.tla]: Multiple blocks in `Buffer` look similar. Would it make sense to introduce an operator to capture those four lines and re-use the operator?

File: spec/fastsync/fastsync_concurrency_with_peers.tla [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+    or do noting
+*)       
+Peer ==
+    \* send peer messages
+    \/ SendPeerMsgs
+    \* do nothing 
+    \/ /\ UNCHANGED buffers
+       /\ UNCHANGED peer
+
+Buffer ==
```
Comment (konnov) [inline spec/fastsync/fastsync_concurrency_with_peers.tla]: The name is kind of misleading. It captures the behavior of multiple buffers, not of one.

File: spec/fastsync/fastsync_concurrency_with_peers.tla [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+       /\ scheduleDemux' = BufRcv(scheduleDemux)
+       /\ UNCHANGED <<receiveDemux, demuxSend, demuxProcess, processDemux, demuxSchedule, peerIncoming, peerOutgoing, peerTurn>>   
+    \/ /\ \E peerID \in PeerIDs : /\ ~Ready(peerIncoming[peerID])
+                                  /\ ~Full(peerIncoming[peerID])
+                                  /\ peerIncoming' = [peerIncoming EXCEPT ![peerID] = BufRcv(@)]
+       /\ UNCHANGED <<receiveDemux, demuxSend, demuxProcess, processDemux, demuxSchedule, scheduleDemux, peerOutgoing, peerTurn>>
+    \/ /\ \E peerID \in PeerIDs : /\ ~Ready(peerOutgoing[peerID])
+                                  /\ ~Full(peerOutgoing[peerID])
+                                  /\ peerOutgoing' = [peerOutgoing EXCEPT ![peerID] = BufRcv(@)]
+       /\ UNCHANGED <<receiveDemux, demuxSend, demuxProcess, processDemux, demuxSchedule, scheduleDemux, peerIncoming, peerTurn>>
```
Comment (konnov) [inline spec/fastsync/fastsync_concurrency_with_peers.tla]: Should we separate internal communication from external in separate operators

File: spec/fastsync/fastsync_concurrency_with_peers.tla [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+       \/ Buffer 
+    /\ turn' \in Turns  
+
+Fairness == 
+    /\ WF_vars(Buffer)
+
+Spec == Init /\ [][Next]_vars /\ Fairness
+  
+
+GoodState ==
```
Comment (konnov) [inline spec/fastsync/fastsync_concurrency_with_peers.tla]: Shall we call it `Invariant`?

**Review by konnov — APPROVED** [review]
The specification looks very nice. Shall we try to check it with apalache, since tlc fails to find a bug?

File: spec/fastsync/fastsync_concurrency_with_peers.tla (hunk 3 above)
Comment (istoilkovska) [reply spec/fastsync/fastsync_concurrency_with_peers.tla]: Not sure what you mean. This is an internal operation of the peerOutgoing buffers (now renamed to peerToReceive), where the inMsg is added to the queue if the buffer is not ready and not full.

**Review by ancazamfir — COMMENTED** [review]
This looks very nice and I learnt a lot reading it :)
I have a doubt on the use of inMsg for buffer. I believe a better model is to read a msg from a queue in inMsg (drain the ingress queue), then if the routine tries to write a message in an outgoing buffer that is full then that routine has to suspend and it will be blocked until the outgoing buffer is drained by the routine at the other end. With this change I reduced your code to 3 and then 2 processes (e.g. having a Demux and a Receive) and I see the  GoodState being violated (which is what we/I hoped for). Maybe we can discuss this a bit.

**Comment by ancazamfir** [comment]
This is one counterexample given by TLC that I tried to illustrate here. The little squares are the memory where we read a message from the inbound queue. Queue size is 1 in this case so when we read a message the queue becomes empty. Circles are routines, greyed one is the one running and when red-lined it is blocked. The D routine injects messages (requests), R responds. It's a simple example but I believe the more complex fastsync architecture with blocking queues should show the same problem.

Untitled Diagram

**Comment by ancazamfir** [comment]
An update on this, after talking to Ilina and thinking on this more, we did a variation of the model where a timer triggered event (e.g. we send a statusRequests every 10 seconds) doesn't run as a separate routine but rather in the same thread of execution as the routine handling the incoming requests (this is the case in v2 where we get a ticker signal into the deumx). 

Concretely, this means that a timer based request cannot go in the outgoing channel D->R if the demux message handling had previously blocked, even if the outgoing channel is not full (like in the last little drawing above). With this new model, the blocking doesn't occur in the simple case. In general, we understand a bit better how to model different concurrency architectures :)

**Comment by konnov** [comment]
If we don't find a way to check the spec for interesting parameters with TLC and Apalache, we can also try to encode this spec in Promela and check it with [Spin](http://spinroot.com/spin/whatispin.html). It was designed for analysis of concurrent systems where processes communicate via bounded channels or shared memory.

**Comment by istoilkovska** [comment]
I've done cleanup and refactoring in the Fastsync concurrency spec, and I've added a README file. In the version of the spec with bounded and blocking buffers, TLC is able to reach a state where all buffers are full and all routines are blocking at diameter 38, in 35min, for buffer length 1 and one peer.

**Comment by ancazamfir** [comment]
Looked through the counterexample for `GoodState` for `FastsyncConcurrencyBoundedBlocking.tla` and it shows a similar counterexample as I described in https://github.com/informalsystems/verification/pull/6#issuecomment-593154773
But as mentioned in https://github.com/informalsystems/verification/pull/6#issuecomment-593536807 this is possible only when the timer based actions happen from a different context and this not the case in the fastsync implementation.
Concretely I believe the problem is with `BroadcastStatusRequest` (state 22 in the trace*) and `SendRProcessBlock` (state 23 in the trace*) actions that are enabled when the `Demux` is "blocked", i.e. it had previously read a message from an inbound queue and blocked while processing that message due to some outgoing queue being full. In this case (bounded blocking queues) the demux cannot ever run again until that particular outgoing queue is drained by the receiver, and this includes any timeout events.

(not sure if runs will consistently show the same counterexample but check state where `demuxInboundMessage` is not `noMsg` and demux action happens)

**Comment by konnov** [comment]
We could model integer timers with Apalache, but the depth of 38 is usually a bit too much for Apalache. It depends on your spec of course.

**Pull request merged**